### ISSUE TITLE
Handle duplicate coordinates with same/different ENVO annotations in RF training and evaluation

### ISSUE THREAD

**Issue by turbomam** [contextualizer-ai/env-embeddings#54]
## Problem

Multiple BioSample records can have:
- **Same inputs**: Same coordinates + same date → **identical Google Earth embeddings**
- **Different outputs**: Different ENVO triad annotations

This creates conflicts for supervised learning.

### Case 1: Same Location, Same ENVO (Technical Replicates)

```
Sample A: (35.7°N, 139.8°E, 2020) → marine biome | coastal zone | sea water
Sample B: (35.7°N, 139.8°E, 2020) → marine biome | coastal zone | sea water
GE embedding: [0.15, -0.04, ..., 0.08] (identical)
```

**Effect**: 
- Inflates dataset size (pseudo-replication)
- Over-represents common locations in training
- Test set leakage if A in train, B in test (they're the same!)

### Case 2: Same Location, Different ENVO (Annotation Disagreement)

```
Sample A: (35.7°N, 139.8°E, 2020) → marine biome | coastal zone | sea water
Sample C: (35.7°N, 139.8°E, 2020) → terrestrial biome | urban | soil
GE embedding: [0.15, -0.04, ..., 0.08] (identical)
```

**Effect**:
- Creates impossible learning task: X → Y₁ AND X → Y₂ where Y₁ ≠ Y₂
- Random Forest will split votes, reduce confidence
- Indicates metadata quality issue (one annotation is wrong)

### Analysis Tasks

1. **Count duplicates**:
```python
# Group by GE embedding (or coordinate + date)
duplicates = df.groupby(['latitude', 'longitude', 'collection_date']).size()
n_duplicate_coords = (duplicates > 1).sum()
n_duplicate_samples = duplicates.sum() - len(duplicates)

print(f"Unique coordinates: {len(duplicates)}")
print(f"Coordinates with duplicates: {n_duplicate_coords}")
print(f"Total duplicate samples: {n_duplicate_samples}")
print(f"Duplication rate: {n_duplicate_samples/len(df)*100:.1f}%")
```

2. **Categorize duplicates**:
```python
duplicate_groups = df[df.duplicated(['latitude', 'longitude', 'collection_date'], keep=False)]

# Same ENVO vs different ENVO
for coord_group in duplicate_groups.groupby(['latitude', 'longitude', 'collection_date']):
    triads = coord_group[['env_broad_scale', 'env_local_scale', 'env_medium']]
    
    if triads.nunique().sum() == 3:
        # All samples have same triad
        print("Technical replicate")
    else:
        # Samples have different triads
        print("Annotation conflict!")
        print(triads.value_counts())
```

3. **Impact on RF training**:
```python
# How many conflicting pairs exist?
conflicting_pairs = []
for (lat, lon, date), group in duplicate_groups.groupby(['latitude', 'longitude', 'collection_date']):
    if len(group['env_local_scale'].unique()) > 1:
        conflicting_pairs.append({
            'coord': (lat, lon, date),
            'n_samples': len(group),
            'envo_values': group['env_local_scale'].unique(),
            'samples': group['accession'].tolist()
        })

print(f"Coordinates with conflicting annotations: {len(conflicting_pairs)}")
```

### Option 1: Deduplicate by Coordinate (Conservative)

Keep only **one sample per unique coordinate**:

```python
# Keep first sample per coordinate
df_deduped = df.drop_duplicates(subset=['latitude', 'longitude', 'collection_date'], keep='first')

# Loss of data but clean training
print(f"Original: {len(df)} → Deduplicated: {len(df_deduped)}")
```

**Pros**: Clean, no conflicts, no pseudo-replication
**Cons**: Loses data, arbitrary choice of which to keep

### Option 2: Majority Vote for Conflicts

For coordinates with multiple annotations, use majority vote:

```python
def resolve_conflicts(group):
    """For duplicate coordinates, pick most common ENVO triad."""
    broad_mode = group['env_broad_scale'].mode()[0]
    local_mode = group['env_local_scale'].mode()[0]
    medium_mode = group['env_medium'].mode()[0]
    
    # Return one representative sample with consensus triad
    rep = group.iloc[0].copy()
    rep['env_broad_scale'] = broad_mode
    rep['env_local_scale'] = local_mode
    rep['env_medium'] = medium_mode
    return rep

df_resolved = df.groupby(['latitude', 'longitude', 'collection_date']).apply(resolve_conflicts)
```

**Pros**: Uses all data to determine "correct" annotation
**Cons**: May hide real biological variation (e.g., seasonal change at same location)

### Option 3: Weight Duplicates Lower

Keep all samples but downweight duplicates in RF training:

```python
# Calculate sample weights (inverse of duplicate count)
coord_counts = df.groupby(['latitude', 'longitude', 'collection_date']).transform('size')
sample_weights = 1.0 / coord_counts

# Train with weights
rf_classifier.fit(X_train, y_train, sample_weight=sample_weights[train_indices])
```

**Pros**: Keeps all data, reduces pseudo-replication bias
**Cons**: Doesn't resolve conflicts, RF still confused by X → Y₁ AND X → Y₂

### Option 4: Flag and Exclude Conflicts

Remove only coordinates with **disagreement**, keep technical replicates:

```python
# Identify conflicting coordinates
conflicting_coords = []
for (lat, lon, date), group in df.groupby(['latitude', 'longitude', 'collection_date']):
    if len(group['env_local_scale'].unique()) > 1:
        conflicting_coords.append((lat, lon, date))

# Exclude conflicting coordinates
df_clean = df[~df.set_index(['latitude', 'longitude', 'collection_date']).index.isin(conflicting_coords)]
```

**Pros**: Keeps clean replicates, removes problematic conflicts
**Cons**: Loses potentially interesting cases (could be metadata errors to fix)

## Recommended Approach

**Two-phase strategy**:

1. **For Issue #49 (initial RF training)**:
   - Use **Option 1 (deduplicate)** for clean baseline
   - Document how many samples removed
   - Report results as "unique coordinates only"

2. **Follow-up analysis** (separate issue or notebook section):
   - Quantify duplicate problem (as above)
   - Try **Option 4 (exclude conflicts only)** to retain technical replicates
   - Compare results: deduplicated vs conflict-excluded
   - Flag conflicting coordinates for manual review (metadata QC)

## Implementation Tasks

- [ ] Add deduplication section to RF notebook before training
- [ ] Report duplicate statistics:
  - [ ] Total duplicates
  - [ ] Technical replicates (same ENVO)
  - [ ] Conflicts (different ENVO)
- [ ] Implement Option 1 (deduplicate) for #49 baseline
- [ ] Document methodology in notebook
- [ ] Save list of conflicting coordinates for manual review
- [ ] Optional: Compare Option 1 vs Option 4 performance

## Expected Findings

**Hypothesis**: 
- NMDC (curated): Low duplication, few conflicts (~1-5%)
- NCBI (uncurated): High duplication, many conflicts (~10-20%)

If conflicts are rare → Option 1 sufficient
If conflicts are common → Need Option 4 or manual review

## Related

- #49 - RF training (needs this for clean data)
- similarity_analysis.ipynb - Already filters "degenerate pairs" (similarity > 0.95)
- Issue about metadata quality (to be created)

## Priority

**High** - Affects RF training quality, should be addressed in #49 before training starts.

## Notes

**Train/Test Split**: Even with deduplication, ensure no coordinate leakage:
- All samples from same coordinate must be in SAME split (train or test)
- Use `GroupShuffleSplit` or custom splitter:

```python
from sklearn.model_selection import GroupShuffleSplit

# Group by coordinate
groups = df['latitude'].astype(str) + '_' + df['longitude'].astype(str)

splitter = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
train_idx, test_idx = next(splitter.split(X, y, groups=groups))
```
